src/SFA.DAS.Payments.ServiceFabric.Core/ServiceBusBatchCommunicationListener.cs

```diff
@@ -0,0 +1,315 @@
+                await Task.WhenAll(
+                    Listen(cancellationToken),
+                    Listen(cancellationToken),
+                    Listen(cancellationToken)
+                    );
+            }
+            catch (Exception ex)
+            {
+                logger.LogFatal($"Encountered fatal error. Error: {ex.Message}", ex);
+            }
+        }
+
+        private async Task Listen(CancellationToken cancellationToken)
+        {
+            var connection = new ServiceBusConnection(connectionString);
+            var messageReceiver = new MessageReceiver(connection, EndpointName, ReceiveMode.PeekLock, RetryPolicy.Default);
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        //var pipeLineStopwatch = Stopwatch.StartNew();
+                        var messages = new List<Message>();
+                        for (var i = 0; i < 10 && messages.Count <= 200; i++)
+                        {
+                            cancellationToken.ThrowIfCancellationRequested();
+                            var receivedMessages = await messageReceiver.ReceiveAsync(200, TimeSpan.FromSeconds(2))
+                                .ConfigureAwait(false);
+                            if (receivedMessages == null || !receivedMessages.Any())
+                                break;
+                            messages.AddRange(receivedMessages);
+                        }
+
+                        if (!messages.Any())
+                        {
+                            await Task.Delay(2000, cancellationToken);
+                            continue;
+                        }
+
+                        var groupedMessages = new Dictionary<Type, List<(string, int, object)>>();
+                        foreach (var message in messages)
+                        {
+                            cancellationToken.ThrowIfCancellationRequested();
+                            try
+                            {
+                                var applicationMessage = DeserializeMessage(message);
+                                var key = applicationMessage.GetType();
+                                var applicationMessages = groupedMessages.ContainsKey(key)
+                                    ? groupedMessages[key]
+                                    : groupedMessages[key] = new List<(string, int, object)>();
+                                applicationMessages.Add((message.SystemProperties.LockToken, message.SystemProperties.DeliveryCount, applicationMessage));
+
+                            }
+                            catch (Exception e)
+                            {
+                                logger.LogError($"Error deserialising the message. Error: {e.Message}", e);
+                                //TODO: should use the error queue instead of dead letter queue
```

Comment:
Again, probably a carry over from the existing codebase. Is there a ticket for this work?

Reply:
See above for the legacy code comment. This one is from 2019.